plugin/plugin.go

```diff
@@ -141,11 +132,186 @@ func (p *AzurePlugin) OnUpdateSet(_ context.Context, req *pb.OnUpdateSetRequest)
 }
 
 func (p *AzurePlugin) ListHosts(ctx context.Context, req *pb.ListHostsRequest) (*pb.ListHostsResponse, error) {
+	startTime := time.Now()
+	defer func() {
+		fmt.Printf("ListHosts completed in %v\n", time.Since(startTime))
+	}()
+
 	if len(req.GetSets()) == 0 {
-		// Nothing to fetch
 		return &pb.ListHostsResponse{}, nil
 	}
-	catalog := req.GetCatalog()
+
+	// Initialize Azure clients and authorization
+	clientStart := time.Now()
+	clients, err := p.initializeAzureResources(req.GetCatalog(), req.GetPersisted().GetSecrets())
+	if err != nil {
+		return nil, err
+	}
+	fmt.Printf("Azure clients initialized in %v\n", time.Since(clientStart))
+
+	// Find matching resources and map them to sets
+	findStart := time.Now()
+	vmResources, vmssResources, resourceToSetMap, err := p.findMatchingResources(ctx, req.GetSets(), clients.resClient)
+	if err != nil {
+		return nil, err
+	}
+	fmt.Printf("Resources discovered in %v (Found %d VMs, %d VMSS)\n",
+		time.Since(findStart),
+		len(vmResources),
+		len(vmssResources))
+
+	// Create mutex and errgroup for parallel processing
+	var mu sync.Mutex
+	vmToNetworkMap := make(map[string]networkInfo)
+	g, ctx := errgroup.WithContext(ctx)
+
+	processingStart := time.Now()
+	// Process standard VMs in parallel if they exist
+	if len(vmResources) > 0 {
+		g.Go(func() error {
+			vmStart := time.Now()
+			vmNetworkMap := make(map[string]networkInfo)
+
+			if err := p.processStandardVMs(ctx, vmResources, clients, vmNetworkMap); err != nil {
```

Comment:
I'd prefer if this function returned the `vmNetworkMap` instead of writing to an argument that was passed in
```suggestion
			vmNetworkMap, err := p.processStandardVMs(ctx, vmResources, clients, vmNetworkMap)
            if err != nil {
```